File: Data Structures & Algorithms/course-schedule/submission-0.java
class Solution {
    public boolean isCyclic(int node, Map<Integer, List<Integer>> adjList, Set<Integer> visited) {
        if (visited.contains(node)) {
            return true;
        }
        if (adjList.getOrDefault(node, new ArrayList<>()).isEmpty()) return false;

        visited.add(node);

        for (int neighbor: adjList.get(node)) {
            if (isCyclic(neighbor, adjList, visited)) {
                return true;
            }
        }

        visited.remove(node);

        return false;
    }

    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> adjList = new HashMap();

        for (int[] course: prerequisites) {
            int u = course[1];
            int v = course[0];

            if (adjList.containsKey(v)) {
                var neighbors = adjList.get(v);
                neighbors.add(u);
                adjList.put(v, neighbors);
            } else {
                var neighbors = new ArrayList();
                neighbors.add(u);
                adjList.put(v, neighbors);
            }
        }

        for (int i = 0; i < numCourses; i++) {
            if (isCyclic(i, adjList, new HashSet())) return false;
        }
        return true;

    }
}

File: Data Structures & Algorithms/count-paths/submission-0.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        cnt = 0
        visited = set()
        def dfs(r, c):
            nonlocal cnt
            if r < 0 or r >= m or c < 0 or c >= n:
                return

            if r == m - 1 and c == n - 1:
                cnt += 1
            
            if (r,c) in visited:
                return
            
            visited.add((r, c))

            dfs(r, c + 1)
            dfs(r + 1, c)

            visited.remove((r, c))
        
        dfs(0, 0)
        return cnt
